### PR TITLE
Ambit

### PR DESCRIPTION
**Pull request by loriab**
## Description
Get Ambit back

## Todos
Notable points that this PR has either accomplished or will accomplish.
* **Developer Interest**
  - [x] This includes some changes I salvaged from the old clean-python branch.
  - [x] Ambit builds with `ENABLE_ambit` as directed from the `external/upstream/ambit/CM`.
  - [x] Ambit _does not_ link into Psi4, as Psi4 doesn't use it. So it's more like Psi4's build system carries it along.
  - [x] The ambit plugin template _is_ in working build order (if you check out the plugin manually; the skeleton plugins are _not_ in working order). It detects the ambit package and links it in. It _doesn't_ however run. See error below that I bet @jturney can patch up in no time.
  - [x] This is associated with branch `cmakelab2` in the main Jet ambit repo. If you haven't already hunted it down, below is the way to work on psi4 repo and ambit repo locally simultaneously.

```
            #GIT_REPOSITORY https://github.com/jturney/ambit
            #GIT_TAG cmakelab2
            DOWNLOAD_COMMAND ""
            SOURCE_DIR "/path/to/ambit-directclone"
```

* **User-Facing for Release Notes**

## Questions
- [x]  This is not ambit in full working order. This is ambit reintegrated to the point that ambit people can patch it up a bit and return it to full working order. 
- [x] @jturney, I didn't renovate the ambit CMake like on some other external projects. If you've a mind to do so, CheMPS2 is quite similar to Ambit in build artifacts and [my CheMPS2 branch config6](https://github.com/loriab/CheMPS2/tree/config6) is current state of the art.

```
Attention! This SCF may be density-fitted.
Traceback (most recent call last):
  File "/theoryfs2/ds/cdsgroup/psi4-compile/hrw-temp/objdir27py35/stage/theoryfs2/ds/cdsgroup/psi4-compile/hrw-temp/install-psi4/bin/psi4", line 223, in <module>
    exec(content)
  File "<string>", line 76, in <module>
  File "/theoryfs2/ds/cdsgroup/psi4-compile/hrw-temp/objdir27py35/stage/theoryfs2/ds/cdsgroup/psi4-compile/hrw-temp/install-psi4/lib//psi4/driver/driver.py", line 447, in energy
    wfn = procedures['energy'][lowername](lowername, molecule=molecule, **kwargs)
  File "./../skeletonambit/pymodule.py", line 65, in run_skeletonambit
    skeletonambit_wfn = psi4.core.plugin('skeletonambit.so', ref_wfn)

RuntimeError: ambit::Tensor::build: Ambit has not been initialized.
```

## Status
- [x]  Ready to go

### CODE REVIEW

**Comment by loriab** [comment]
FYI, @jturney, this is up-to-date again and ready for ambit-side fixing. Note-to-self: branch `ambit2` locally.

**Comment by fevangelista** [comment]
@loriab: the following patches to `psi4/share/psi4/plugin/ambit/CMakeLists.txt.template` should fix the above problem. Can you try it when you have a chance?

line 54
```
extern "C"
SharedWavefunction @plugin@(SharedWavefunction ref_wfn, Options &options)
{
    using namespace ambit;

    ambit::initialize(); // <== add
```

line 181:
```
    //B.print();

    ambit::finalize();  // <== add

    return ref_wfn;
```

File: psi4/driver/p4util/util.py [hunk 1]
```diff
@@ -382,6 +382,34 @@ def compare_cubes(expected, computed, label):
...
+            break
+
+    if(failed):
+        core.print_out("The computed vector\n")
+        computed.print_out()
+        core.print_out("The reference vector\n")
+        expected.print_out()
+        message = ("\t%s: computed list does not match expected list." % (label, computed, expected))
+        raise TestComparisonError(message)
+    success(label)
```
Comment (dgasmith) [inline psi4/driver/p4util/util.py]: Any reason not to use the `compare_arrays` here? NumPy can handle lists easily.

Comment (loriab) [inline psi4/driver/p4util/util.py]: I think I used this branch to salvage any still-good parts of the old https://github.com/psi4/psi4private/tree/clean_python branch. It's not used anywhere that I can see. I'm going to rebase and push this branch soon. How about if I comment it out, then any future need we'll try compare_arrays, and barring success, will uncomment this?

Comment (dgasmith) [inline psi4/driver/p4util/util.py]: Sounds good.

File: psi4/share/psi4/plugin/ambit/__init__.py.template [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+#
+# Psi4: an open-source quantum chemistry software package
+#
+# Copyright (c) 2007-2017 The Psi4 Developers.
+#
+# The copyrights for code used from other parties are included in
+# the corresponding files.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
```
Comment (loriab) [inline psi4/share/psi4/plugin/ambit/__init__.py.template]: Ack. And I'll fix the licensing.

**Pull request merged**